File: src/file_manager.py
"""File management utilities for the Scryfall MCP server."""

import os
import uuid
import mimetypes
from pathlib import Path
from typing import Optional, Dict, Tuple
from db_manager import CardDatabase


class FileManager:
    """Manages file operations and security for the MCP server."""
    
    def __init__(self, db: Optional[CardDatabase] = None):
        """
        Initialize the file manager.
        
        Args:
            db: Optional CardDatabase instance (will create one if not provided)
        """
        self.db = db
        self._owns_db = db is None
        if self._owns_db:
            self.db = CardDatabase()
    
    def generate_file_id(self) -> str:
        """Generate a unique file ID."""
        return str(uuid.uuid4())
    
    def get_file_path(self, file_id: str) -> Optional[Path]:
        """
        Get the file path for a given file ID.
        
        Args:
            file_id: The unique file ID
            
        Returns:
            Path object if found, None otherwise
        """
        file_path = self.db.get_file_path_by_id(file_id)
        if file_path:
            path = Path(file_path)
            # Security check: ensure the path exists and is a file
            if path.exists() and path.is_file():
                return path
        return None
    
    def get_mime_type(self, file_path: Path) -> str:
        """
        Determine the MIME type of a file.
        
        Args:
            file_path: Path to the file
            
        Returns:
            MIME type string
        """
        mime_type, _ = mimetypes.guess_type(str(file_path))
        if mime_type:
            return mime_type
        
        # Default MIME types for common extensions
        extension = file_path.suffix.lower()
        mime_map = {
            '.jpg': 'image/jpeg',
            '.jpeg': 'image/jpeg',
            '.png': 'image/png',
            '.gif': 'image/gif',
            '.webp': 'image/webp',
            '.json': 'application/json',
            '.txt': 'text/plain'
        }
        
        return mime_map.get(extension, 'application/octet-stream')
    
    def read_file_content(self, file_id: str) -> Optional[Tuple[bytes, str]]:
        """
        Read file content by file ID.
        
        Args:
            file_id: The unique file ID
            
        Returns:
            Tuple of (file content, mime type) or None if not found
        """
        file_path = self.get_file_path(file_id)
        if not file_path:
            return None
        
        try:
            with open(file_path, 'rb') as f:
                content = f.read()
            mime_type = self.get_mime_type(file_path)
            return content, mime_type
        except (OSError, IOError):
            return None
    
    def register_file(self, file_path: Path, card_name: str, 
                     card_id: Optional[str] = None,
                     set_code: Optional[str] = None,
                     image_url: Optional[str] = None) -> str:
        """
        Register a file in the database and return its file ID.
        
        Args:
            file_path: Path to the file
            card_name: Name of the card
            card_id: Scryfall card ID (optional)
            set_code: Set code (optional)
            image_url: Image URL (optional)
            
        Returns:
            The generated file ID
        """
        file_id = self.generate_file_id()
        self.db.add_card(
            card_name=card_name,
            filename=str(file_path),
            card_id=card_id,
            set_code=set_code,
            image_url=image_url,
            file_id=file_id
        )
        return file_id
    
    def cleanup_temp_files(self, directory: Path, max_age_hours: int = 24) -> int:
        """
        Clean up old temporary files.
        
        Args:
            directory: Directory to clean
            max_age_hours: Maximum age in hours for files to keep
            
        Returns:
            Number of files deleted
        """
        import time
        
        if not directory.exists():
            return 0
        
        deleted_count = 0
        current_time = time.time()
        max_age_seconds = max_age_hours * 3600
        
        for file_path in directory.rglob('*'):
            if file_path.is_file():
                file_age = current_time - file_path.stat().st_mtime
                if file_age > max_age_seconds:
                    try:
                        file_path.unlink()
                        deleted_count += 1
                    except (OSError, PermissionError):
                        pass
        
        return deleted_count
    
    def validate_resource_access(self, resource_type: str, file_id: str) -> bool:
        """
        Validate that a resource access is allowed.
        
        Args:
            resource_type: Type of resource (e.g., 'card', 'art', 'metadata')
            file_id: The file ID to access
            
        Returns:
            True if access is allowed, False otherwise
        """
        # Ensure file_id is a valid UUID format
        try:
            uuid.UUID(file_id)
        except ValueError:
            return False
        
        # Check if the file exists in the database
        return self.get_file_path(file_id) is not None
    
    def get_resource_info(self, file_id: str) -> Optional[Dict[str, str]]:
        """
        Get information about a resource.
        
        Args:
            file_id: The file ID
            
        Returns:
            Dictionary with resource information or None
        """
        card_info = self.db.get_card_by_file_id(file_id)
        if not card_info:
            return None
        
        file_path = Path(card_info['filename'])
        
        return {
            'card_name': card_info['card_name'],
            'filename': file_path.name,
            'mime_type': self.get_mime_type(file_path),
            'download_date': card_info['download_date'],
            'set_code': card_info['set_code'] or '',
            'card_id': card_info['card_id'] or ''
        }
    
    def close(self):
        """Close the database connection if we own it."""
        if self._owns_db and self.db:
            self.db.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

File: src/fileManager.ts
/**
 * File management utilities for the Scryfall MCP server.
 */

import { readFile, stat, unlink, readdir } from 'node:fs/promises';
import { existsSync, statSync } from 'node:fs';
import { join, extname, basename } from 'node:path';
import { lookup } from 'mime-types';
import { v4 as uuidv4, validate as isValidUuid } from 'uuid';
import { CardDatabase, createCardDatabase } from './database.js';
import { logger } from './logger.js';

export interface ResourceInfo {
  card_name: string;
  filename: string;
  mime_type: string;
  download_date: string;
  set_code: string;
  card_id: string;
}

export class FileManager {
  private db: CardDatabase | null = null;
  private ownsDb: boolean = false;

  constructor(db?: CardDatabase) {
    this.db = db || null;
    this.ownsDb = !db;
  }

  /**
   * Initialize the file manager with database connection.
   */
  async init(): Promise<void> {
    if (!this.db) {
      this.db = await createCardDatabase();
    }
  }

  /**
   * Generate a unique file ID.
   */
  generateFileId(): string {
    return uuidv4();
  }

  /**
   * Get the file path for a given file ID.
   */
  async getFilePath(fileId: string): Promise<string | undefined> {
    if (!this.db) {
      await this.init();
    }

    const filePath = this.db!.getFilePathById(fileId);
    if (filePath) {
      // Security check: ensure the path exists and is a file
      if (existsSync(filePath)) {
        const stats = statSync(filePath);
        if (stats.isFile()) {
          return filePath;
        }
      }
    }
    return undefined;
  }

  /**
   * Determine the MIME type of a file.
   */
  getMimeType(filePath: string): string {
    const mimeType = lookup(filePath);
    if (mimeType) {
      return mimeType;
    }

    // Default MIME types for common extensions
    const extension = extname(filePath).toLowerCase();
    const mimeMap: Record<string, string> = {
      '.jpg': 'image/jpeg',
      '.jpeg': 'image/jpeg',
      '.png': 'image/png',
      '.gif': 'image/gif',
      '.webp': 'image/webp',
      '.json': 'application/json',
      '.txt': 'text/plain'
    };

    return mimeMap[extension] || 'application/octet-stream';
  }

  /**
   * Read file content by file ID with enhanced security and error handling.
   */
  async readFileContent(fileId: string): Promise<[Buffer, string] | undefined> {
    try {
      // Validate access first
      const isAccessAllowed = await this.validateResourceAccess('download/card', fileId);
      if (!isAccessAllowed) {
        logger.warn(`Access denied for file ID: ${fileId}`);
        return undefined;
      }

      const filePath = await this.getFilePath(fileId);
      if (!filePath) {
        logger.warn(`File path not found for file ID: ${fileId}`);
        return undefined;
      }

      // Double-check file exists and is readable
      try {
        await stat(filePath);
      } catch (statError: any) {
        logger.error(`File not accessible: ${filePath} - ${statError.message}`);
        return undefined;
      }

      // Read file content with size validation
      const stats = await stat(filePath);
      const maxFileSize = 50 * 1024 * 1024; // 50MB limit
      if (stats.size > maxFileSize) {
        logger.error(`File size exceeds limit: ${stats.size} bytes for ${filePath}`);
        return undefined;
      }

      const content = await readFile(filePath);
      const mimeType = this.getMimeType(filePath);
      
      logger.info(`Successfully read file: ${basename(filePath)} (${content.length} bytes, ${mimeType})`);
      return [content, mimeType];
      
    } catch (error: any) {
      logger.error(`Failed to read file content for ${fileId}: ${error.message}`);
      return undefined;
    }
  }

  /**
   * Register a file in the database and return its file ID with enhanced validation.
   */
  async registerFile(
    filePath: string,
    cardName: string,
    cardId?: string,
    setCode?: string,
    imageUrl?: string
  ): Promise<string> {
    try {
      if (!this.db) {
        await this.init();
      }

      // Validate inputs
      if (!filePath || !cardName) {
        throw new Error('File path and card name are required');
      }

      // Verify file exists before registering
      if (!existsSync(filePath)) {
        throw new Error(`File does not exist: ${filePath}`);
      }

      // Validate file is actually a file (not directory)
      const stats = statSync(filePath);
      if (!stats.isFile()) {
        throw new Error(`Path is not a file: ${filePath}`);
      }

      // Validate file extension
      const extension = extname(filePath).toLowerCase();
      const validExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.json', '.txt'];
      if (!validExtensions.includes(extension)) {
        logger.warn(`Registering file with unusual extension: ${extension}`);
      }

      const fileId = this.generateFileId();
      
      try {
        this.db!.addCard(
          cardName,
          filePath,
          cardId,
          setCode,
          imageUrl,
          fileId
        );
        
        logger.info(`File registered successfully: ${cardName} -> ${fileId} (${basename(filePath)})`);
        return fileId;
        
      } catch (dbError: any) {
        logger.error(`Database error during file registration: ${dbError.message}`);
        throw new Error(`Failed to register file in database: ${dbError.message}`);
      }
      
    } catch (error: any) {
      logger.error(`Failed to register file ${filePath}: ${error.message}`);
      throw error;
    }
  }

  /**
   * Clean up old temporary files.
   */
  async cleanupTempFiles(directory: string, maxAgeHours: number = 24): Promise<number> {
    if (!existsSync(directory)) {
      return 0;
    }

    let deletedCount = 0;
    const currentTime = Date.now();
    const maxAgeMs = maxAgeHours * 3600 * 1000;

    try {
      const files = await this.walkDirectory(directory);
      
      for (const filePath of files) {
        try {
          const stats = await stat(filePath);
          const fileAge = currentTime - stats.mtime.getTime();
          
          if (fileAge > maxAgeMs) {
            await unlink(filePath);
            deletedCount++;
          }
        } catch (error) {
          // Ignore errors for individual files
        }
      }
    } catch (error) {
      logger.error('Failed to cleanup temp files:', error);
    }

    return deletedCount;
  }

  /**
   * Recursively walk a directory and return all file paths.
   */
  private async walkDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    
    try {
      const entries = await readdir(directory, { withFileTypes: true });
      
      for (const entry of entries) {
        const fullPath = join(directory, entry.name);
        
        if (entry.isFile()) {
          files.push(fullPath);
        } else if (entry.isDirectory()) {
          const subFiles = await this.walkDirectory(fullPath);
          files.push(...subFiles);
        }
      }
    } catch (error) {
      // Ignore errors for individual directories
    }
    
    return files;
  }

  /**
   * Validate that a resource access is allowed with comprehensive security checks.
   */
  async validateResourceAccess(resourceType: string, fileId: string): Promise<boolean> {
    try {
      // Ensure file_id is a valid UUID format
      if (!isValidUuid(fileId)) {
        logger.warn(`Invalid UUID format for file access: ${fileId}`);
        return false;
      }

      // Validate resource type
      const allowedResourceTypes = ['download/card', 'download/art', 'download/metadata'];
      if (!allowedResourceTypes.includes(resourceType)) {
        logger.warn(`Invalid resource type: ${resourceType}`);
        return false;
      }

      // Check if file exists in database
      if (!this.db) {
        await this.init();
      }

      const cardInfo = this.db!.getCardByFileId(fileId);
      if (!cardInfo) {
        logger.warn(`File ID not found in database: ${fileId}`);
        return false;
      }

      // Check if physical file exists
      const filePath = cardInfo.filename;
      if (!existsSync(filePath)) {
        logger.warn(`Physical file does not exist: ${filePath}`);
        return false;
      }

      // Verify file is within allowed directories
      const resolvedPath = require('path').resolve(filePath);
      const allowedDirs = [
        '/tmp/scryfall_downloads',
        '/tmp/scryfall_fallback',
        process.env['SCRYFALL_DATA_DIR'],
        require('os').tmpdir(),
        require('path').join(require('os').homedir(), '.local')
      ].filter(Boolean);

      const isInAllowedDir = allowedDirs.some(dir => {
        if (!dir) return false;
        const resolvedDir = require('path').resolve(dir);
        return resolvedPath.startsWith(resolvedDir);
      });

      if (!isInAllowedDir) {
        logger.warn(`File path outside allowed directories: ${resolvedPath}`);
        return false;
      }

      // Check file size limits (reasonable limits for card images)
      const stats = statSync(filePath);
      const maxFileSize = 50 * 1024 * 1024; // 50MB limit
      if (stats.size > maxFileSize) {
        logger.warn(`File size exceeds limit: ${stats.size} bytes`);
        return false;
      }

      // Validate file extension for expected content types
      const extension = extname(filePath).toLowerCase();
      const validExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.json', '.txt'];
      if (!validExtensions.includes(extension)) {
        logger.warn(`Invalid file extension: ${extension}`);
        return false;
      }

      return true;
      
    } catch (error: any) {
      logger.error(`Error validating resource access: ${error.message}`);
      return false;
    }
  }

  /**
   * Get comprehensive information about a resource, matching Python version functionality.
   */
  async getResourceInfo(fileId: string): Promise<ResourceInfo | undefined> {
    try {
      if (!this.db) {
        await this.init();
      }

      // Validate file ID format
      if (!isValidUuid(fileId)) {
        logger.warn(`Invalid UUID format for resource info: ${fileId}`);
        return undefined;
      }

      const cardInfo = this.db!.getCardByFileId(fileId);
      if (!cardInfo) {
        logger.warn(`Card info not found for file ID: ${fileId}`);
        return undefined;
      }

      // Verify physical file exists
      const filePath = cardInfo.filename;
      if (!existsSync(filePath)) {
        logger.warn(`Physical file missing for resource info: ${filePath}`);
        return undefined;
      }

      // Get file stats for additional metadata
      let fileStats;
      try {
        fileStats = await stat(filePath);
      } catch (statError: any) {
        logger.error(`Failed to get file stats: ${statError.message}`);
        return undefined;
      }

      const resourceInfo: ResourceInfo = {
        card_name: cardInfo.card_name,
        filename: basename(cardInfo.filename),
        mime_type: this.getMimeType(cardInfo.filename),
        download_date: cardInfo.download_date,
        set_code: cardInfo.set_code || '',
        card_id: cardInfo.card_id || ''
      };

      // Additional metadata that might be useful
      logger.info(`Resource info retrieved for ${fileId}: ${resourceInfo.card_name} (${resourceInfo.filename})`);
      
      return resourceInfo;
      
    } catch (error: any) {
      logger.error(`Failed to get resource info for ${fileId}: ${error.message}`);
      return undefined;
    }
  }

  /**
   * Close the database connection if we own it.
   */
  close(): void {
    if (this.ownsDb && this.db) {
      this.db.close();
    }
  }
}

/**
 * Create a file manager instance with automatic initialization.
 */
export async function createFileManager(db?: CardDatabase): Promise<FileManager> {
  const fm = new FileManager(db);
  await fm.init();
  return fm;
}